File: src/main/java/org/stt/gui/jfx/TimeTrackingItemCellWithActions.java
package org.stt.gui.jfx;

import com.sun.javafx.geom.BaseBounds;
import com.sun.javafx.geom.transform.BaseTransform;
import com.sun.javafx.jmx.MXNodeAlgorithm;
import com.sun.javafx.jmx.MXNodeAlgorithmContext;
import com.sun.javafx.sg.prism.NGNode;
import javafx.animation.FadeTransition;
import javafx.beans.binding.Bindings;
import javafx.beans.property.DoubleProperty;
import javafx.geometry.Insets;
import javafx.geometry.Pos;
import javafx.scene.Node;
import javafx.scene.control.*;
import javafx.scene.layout.Background;
import javafx.scene.layout.BackgroundFill;
import javafx.scene.layout.BorderPane;
import javafx.scene.layout.HBox;
import javafx.scene.paint.Color;
import javafx.scene.text.Font;
import org.stt.model.TimeTrackingItem;

import java.time.format.DateTimeFormatter;
import java.time.format.FormatStyle;
import java.util.ResourceBundle;
import java.util.function.Predicate;
import java.util.stream.Stream;

import static java.util.Objects.requireNonNull;
import static org.stt.gui.jfx.Glyph.GLYPH_SIZE_MEDIUM;
import static org.stt.gui.jfx.Glyph.glyph;

class TimeTrackingItemCellWithActions extends ListCell<TimeTrackingItem> {
    private static final DateTimeFormatter TIME_FORMATTER = DateTimeFormatter.ofLocalizedTime(FormatStyle.MEDIUM);
    private static final DateTimeFormatter DATE_FORMATTER = DateTimeFormatter.ofLocalizedDate(FormatStyle.FULL);

    private final HBox cellPane = new HBox(2);
    final Button editButton;
    final Button continueButton;
    final Button deleteButton;
    final Button stopButton;
    private final Predicate<TimeTrackingItem> lastItemOfDay;
    private final BorderPane lastItemOnDayPane;

    private final Node newDayNode;
    private final TimeTrackingItemNodes itemNodes;
    private final int buttonIndex;

    TimeTrackingItemCellWithActions(Font fontAwesome,
                                    ResourceBundle localization,
                                    Predicate<TimeTrackingItem> lastItemOfDay,
                                    ActionsHandler actionsHandler,
                                    ActivityTextDisplayProcessor labelToNodeMapper) {
        requireNonNull(fontAwesome);
        requireNonNull(actionsHandler);
        itemNodes = new TimeTrackingItemNodes(labelToNodeMapper, TIME_FORMATTER, fontAwesome, 450, 180, localization);
        editButton = new FramelessButton(glyph(fontAwesome, Glyph.PENCIL, GLYPH_SIZE_MEDIUM));
        continueButton = new FramelessButton(glyph(fontAwesome, Glyph.PLAY_CIRCLE, GLYPH_SIZE_MEDIUM, Color.DARKGREEN));
        deleteButton = new FramelessButton(glyph(fontAwesome, Glyph.TRASH, GLYPH_SIZE_MEDIUM, Color.web("e26868")));
        stopButton = new FramelessButton(glyph(fontAwesome, Glyph.STOP_CIRCLE, GLYPH_SIZE_MEDIUM, Color.GOLDENROD));
        this.lastItemOfDay = requireNonNull(lastItemOfDay);
        setupTooltips(localization);

        continueButton.setOnAction(event -> actionsHandler.continueItem(getItem()));
        editButton.setOnAction(event -> actionsHandler.edit(getItem()));
        deleteButton.setOnAction(event -> actionsHandler.delete(getItem()));
        stopButton.setOnAction(event -> actionsHandler.stop(getItem()));

        itemNodes.appendNodesTo(cellPane.getChildren());

        buttonIndex = cellPane.getChildren().size();

        cellPane.getChildren().addAll(
                continueButton,
                editButton,
                deleteButton);
        cellPane.setAlignment(Pos.CENTER_LEFT);

        DoubleProperty opacityProperty = fadeOnHoverProperty();

        Stream.of(continueButton, editButton, deleteButton, stopButton)
                .map(Node::opacityProperty)
                .forEach(p -> p.bind(opacityProperty));

        lastItemOnDayPane = new BorderPane();

        newDayNode = createDateSubheader(fontAwesome);

        setContentDisplay(ContentDisplay.GRAPHIC_ONLY);
    }

    private DoubleProperty fadeOnHoverProperty() {
        Node placeHolder = new DummyNode();
        placeHolder.setOpacity(0);

        FadeTransition fade = new FadeTransition();
        fade.setFromValue(0);
        fade.setToValue(1);
        fade.setNode(placeHolder);
        cellPane.hoverProperty().addListener((observable, oldValue, newValue) -> {
            fade.setRate(newValue ? 1 : -1);
            fade.play();
        });
        return placeHolder.opacityProperty();
    }

    private Node createDateSubheader(Font fontAwesome) {
        HBox newDayHbox = new HBox(5);
        newDayHbox.setPadding(new Insets(2));
        Label calenderIcon = glyph(fontAwesome, Glyph.CALENDAR);
        calenderIcon.setTextFill(Color.BLACK);
        newDayHbox.getChildren().add(calenderIcon);
        Label dayLabel = new Label();
        dayLabel.setTextFill(Color.BLACK);
        dayLabel.textProperty().bind(Bindings.createStringBinding(
                () -> itemProperty().get() == null ? "" : DATE_FORMATTER.format(itemProperty().get().getStart()),
                itemProperty()));
        newDayHbox.getChildren().add(dayLabel);
        newDayHbox.setBackground(new Background(new BackgroundFill(Color.LIGHTSTEELBLUE, null, null)));
        return newDayHbox;
    }

    protected void setupTooltips(ResourceBundle localization) {
        editButton.setTooltip(new Tooltip(localization
                .getString("itemList.edit")));
        continueButton.setTooltip(new Tooltip(localization
                .getString("itemList.continue")));
        deleteButton.setTooltip(new Tooltip(localization
                .getString("itemList.delete")));
        stopButton.setTooltip(new Tooltip(localization
                .getString("itemList.stop")));
    }

    @Override
    protected void updateItem(TimeTrackingItem item, boolean empty) {
        super.updateItem(item, empty);
        if (empty) {
            setGraphic(null);
        } else {
            cellPane.getChildren().set(buttonIndex, item.getEnd().isPresent() ? continueButton : stopButton);
            itemNodes.setItem(item);
            if (lastItemOfDay.test(item)) {
                setupLastItemOfDayPane();
                setGraphic(lastItemOnDayPane);
            } else {
                setupCellPane();
                setGraphic(cellPane);
            }
        }
    }

    private void setupCellPane() {
        lastItemOnDayPane.setCenter(null);
        lastItemOnDayPane.setTop(null);
    }

    private void setupLastItemOfDayPane() {
        lastItemOnDayPane.setCenter(cellPane);
        lastItemOnDayPane.setTop(newDayNode);
        BorderPane.setMargin(newDayNode, new Insets(10, 0, 10, 0));
    }

    public interface ActionsHandler {
        void continueItem(TimeTrackingItem item);

        void edit(TimeTrackingItem item);

        void delete(TimeTrackingItem item);

        void stop(TimeTrackingItem item);
    }

    private static class DummyNode extends Node {
        @Override
        protected NGNode impl_createPeer() {
            return null;
        }

        @Override
        public BaseBounds impl_computeGeomBounds(BaseBounds bounds, BaseTransform tx) {
            return null;
        }

        @Override
        protected boolean impl_computeContains(double localX, double localY) {
            return false;
        }

        @Override
        public Object impl_processMXNode(MXNodeAlgorithm alg, MXNodeAlgorithmContext ctx) {
            return null;
        }
    }
}

File: src/main/kotlin/org/stt/gui/jfx/TimeTrackingItemCellWithActions.kt
package org.stt.gui.jfx

import javafx.animation.FadeTransition
import javafx.beans.binding.Bindings
import javafx.beans.property.SimpleDoubleProperty
import javafx.geometry.Insets
import javafx.geometry.Pos
import javafx.scene.Node
import javafx.scene.control.*
import javafx.scene.layout.*
import javafx.scene.paint.Color
import javafx.scene.text.Font
import org.stt.gui.jfx.Glyph.Companion.GLYPH_SIZE_MEDIUM
import org.stt.gui.jfx.Glyph.Companion.glyph
import org.stt.model.TimeTrackingItem
import org.stt.submit.SubmitSelectionManager
import org.stt.submit.SubmitStatusTracker
import java.time.format.DateTimeFormatter
import java.time.format.FormatStyle
import java.util.*
import java.util.concurrent.Callable
import java.util.function.Predicate

internal open class TimeTrackingItemCellWithActions(fontAwesome: Font,
                                                    localization: ResourceBundle,
                                                    private val lastItemOfDay: Predicate<TimeTrackingItem>,
                                                    actionsHandler: ActionsHandler,
                                                    labelToNodeMapper: ActivityTextDisplayProcessor,
                                                    private val submitStatusTracker: SubmitStatusTracker,
                                                    private val submitSelectionManager: SubmitSelectionManager,
                                                    private val connectorIdProvider: () -> String?) : ListCell<TimeTrackingItem>() {

    private val cellPane = HBox(2.0)
    val editButton: Button
    val continueButton: Button
    val deleteButton: Button
    val stopButton: Button
    private val submitCheckBox: CheckBox
    private val lastItemOnDayPane: BorderPane

    private val newDayNode: Node
    private val itemNodes: TimeTrackingItemNodes
    private val actions: HBox

    init {
        itemNodes = TimeTrackingItemNodes(labelToNodeMapper, TIME_FORMATTER, fontAwesome, 450, 220, localization)
        editButton = FramelessButton(glyph(fontAwesome, Glyph.PENCIL, GLYPH_SIZE_MEDIUM))
        continueButton = FramelessButton(glyph(fontAwesome, Glyph.PLAY_CIRCLE, GLYPH_SIZE_MEDIUM, Color.DARKGREEN))
        deleteButton = FramelessButton(glyph(fontAwesome, Glyph.TRASH, GLYPH_SIZE_MEDIUM, Color.web("e26868")))
        stopButton = FramelessButton(glyph(fontAwesome, Glyph.STOP_CIRCLE, GLYPH_SIZE_MEDIUM, Color.GOLDENROD))
        submitCheckBox = CheckBox()
        setupTooltips(localization)

        continueButton.setOnAction { actionsHandler.continueItem(item) }
        editButton.setOnAction { actionsHandler.edit(item) }
        deleteButton.setOnAction { actionsHandler.delete(item) }
        stopButton.setOnAction { actionsHandler.stop(item) }

        submitCheckBox.setOnAction {
            val connectorId = connectorIdProvider() ?: return@setOnAction
            val currentItem = item ?: return@setOnAction
            submitSelectionManager.setSelected(connectorId, currentItem, submitCheckBox.isSelected)
        }

        actions = HBox(continueButton, editButton, deleteButton)
        StackPane.setAlignment(actions, Pos.CENTER)
        val timeOrActions = StackPaneWithoutResize()

        val fade = addOpacityFadeTransition(actions);
        bindFadeTransitionToHover(fade, cellPane);

        val one = SimpleDoubleProperty(1.0)
        val timePaneOpacity = one.subtract(Bindings.min(one, actions.opacityProperty().multiply(2)))
        itemNodes.appendNodesToAndWrapTimePane(timeOrActions, cellPane.children)
        itemNodes.bindTimePaneOpacity(timePaneOpacity)
        timeOrActions.children.add(actions)

        cellPane.children.add(submitCheckBox)
        cellPane.alignment = Pos.CENTER_LEFT

        lastItemOnDayPane = BorderPane()

        newDayNode = createDateSubheader(fontAwesome)

        contentDisplay = ContentDisplay.GRAPHIC_ONLY
    }

    private fun addOpacityFadeTransition(node: Node) : FadeTransition {
        node.opacity = 0.0

        val fade = FadeTransition()
        fade.fromValue = 0.0
        fade.toValue = 1.0
        fade.node = node
        cellPane.hoverProperty().addListener { _, _, newValue ->
            fade.rate = (if (newValue) 1 else -1).toDouble()
            fade.play()
        }
        return fade;
    }

    private fun bindFadeTransitionToHover(fade : FadeTransition, node: Node) {
        node.hoverProperty().addListener { _, _, newValue ->
            fade.rate = (if (newValue) 1 else -1).toDouble()
            fade.play()
        }
    }

    private fun createDateSubheader(fontAwesome: Font): Node {
        val newDayHbox = HBox(5.0)
        newDayHbox.padding = Insets(2.0)
        val calenderIcon = glyph(fontAwesome, Glyph.CALENDAR)
        calenderIcon.textFill = Color.BLACK
        newDayHbox.children.add(calenderIcon)
        val dayLabel = Label()
        dayLabel.textFill = Color.BLACK
        dayLabel.textProperty().bind(Bindings.createStringBinding(
                Callable { if (itemProperty().get() == null) "" else DATE_FORMATTER.format(itemProperty().get().start) },
                itemProperty()))
        newDayHbox.children.add(dayLabel)
        newDayHbox.background = Background(BackgroundFill(Color.LIGHTSTEELBLUE, null, null))
        return newDayHbox
    }

    fun setupTooltips(localization: ResourceBundle) {
        editButton.tooltip = Tooltip(localization
                .getString("itemList.edit"))
        continueButton.tooltip = Tooltip(localization
                .getString("itemList.continue"))
        deleteButton.tooltip = Tooltip(localization
                .getString("itemList.delete"))
        stopButton.tooltip = Tooltip(localization
                .getString("itemList.stop"))
    }

    public override fun updateItem(item: TimeTrackingItem?, empty: Boolean) {
        super.updateItem(item, empty)
        if (empty || item == null) {
            graphic = null
        } else {
            actions.children[0] = if (item.end != null) continueButton else stopButton

            val isSubmitted = submitStatusTracker.isSubmitted(item)
            editButton.isDisable = isSubmitted
            deleteButton.isDisable = isSubmitted
            continueButton.isDisable = isSubmitted
            stopButton.isDisable = isSubmitted

            val connectorId = connectorIdProvider()
            val isSubmittedToActive = connectorId?.let { submitStatusTracker.isSubmitted(item, it) } ?: false
            submitCheckBox.isSelected = isSubmittedToActive || (connectorId?.let { submitSelectionManager.isSelected(it, item) } ?: false)
            submitCheckBox.isDisable = isSubmittedToActive

            itemNodes.setItem(item)
            graphic = if (lastItemOfDay.test(item)) {
                setupLastItemOfDayPane()
                lastItemOnDayPane
            } else {
                setupCellPane()
                cellPane
            }
        }
    }

    private fun setupCellPane() {
        lastItemOnDayPane.center = null
        lastItemOnDayPane.top = null
    }

    private fun setupLastItemOfDayPane() {
        lastItemOnDayPane.center = cellPane
        lastItemOnDayPane.top = newDayNode
        BorderPane.setMargin(newDayNode, Insets(10.0, 0.0, 10.0, 0.0))
    }

    interface ActionsHandler {
        fun continueItem(item: TimeTrackingItem)

        fun edit(item: TimeTrackingItem)

        fun delete(item: TimeTrackingItem)

        fun stop(item: TimeTrackingItem)
    }

    companion object {
        private val TIME_FORMATTER = DateTimeFormatter.ofLocalizedTime(FormatStyle.MEDIUM)
        private val DATE_FORMATTER = DateTimeFormatter.ofLocalizedDate(FormatStyle.FULL)
    }
}